File: week12/2/Prob4.cpp
#include <stack>

using namespace std;

class MyQueue {
private:
    stack<int> stack1, stack2;

public:
    MyQueue(){
        // 기본 생성자
    }
    
    void push(int x){
        // push 시 stack1에 데이터를 추가
        stack1.push(x);
    }
    
    int pop(){
        // stack2가 비어 있는 경우에만 stack1의 모든 데이터를 stack2로 옮김
        if(stack2.empty()){
            while(!stack1.empty()){
                stack2.push(stack1.top());
                stack1.pop();
            }
        }
        // stack2의 맨 위의 요소를 제거하고 반환
        int topElement = stack2.top();
        stack2.pop();
        return topElement;
    }
    
    int peek(){
        // pop과 마찬가지로 stack2가 비어 있는 경우 stack1의 모든 데이터를 stack2로 옮김
        if(stack2.empty()){
            while(!stack1.empty()){
                stack2.push(stack1.top());
                stack1.pop();
            }
        }
        // stack2의 맨 위의 요소를 반환
        return stack2.top();
    }
    
    bool empty(){
        // 두 스택이 모두 비어 있으면 큐가 비어 있는 것으로 간주
        return stack1.empty() && stack2.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: week12/2/Prob4.py
class MyQueue:

    def __init__(self):
        # 스택 두 개를 활용하여 queue를 구현 
        self.stack_in = []  # 입력 스택 (push 용도)
        self.stack_out = []  # 출력 스택 (pop, peek 용도)

    def push(self, x: int) -> None:
        self.stack_in.append(x)

    def pop(self) -> int:
        # 출력 스택이 비어 있다면, 입력 스택의 모든 요소를 옮겨 담음
        if not self.stack_out:
            while self.stack_in:
                self.stack_out.append(self.stack_in.pop())
        # 출력 스택의 맨 위 요소를 반환: FIFO 순서를 유지
        return self.stack_out.pop()

    def peek(self) -> int:
        # 출력 스택이 비어 있다면, 입력 스택의 모든 요소를 옮겨 담음
        if not self.stack_out:
            while self.stack_in:
                self.stack_out.append(self.stack_in.pop())
        # 출력 스택의 맨 마지막 요소를 반환: 큐의 front 요소
        return self.stack_out[-1]

    def empty(self) -> bool:
        # 입력 스택과 출력 스택이 모두 비어 있으면 큐는 비어있다고 판단
        return not self.stack_in and not self.stack_out
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()


input = ["MyQueue","push","push","peek","pop","empty"]
inputval = [[],[1],[2],[],[],[]]

result = []

# 테스트 환경 구현, output check 용도
for i in range(len(input)):
    #print(i, input[i], inputval[i])
    if input[i] == "MyQueue":
        obj = MyQueue()
        result.append(None)
    elif input[i] == "push":
        obj.push(inputval[i][0])
        result.append(None)
    elif input[i] == "peek":
        result.append(obj.peek())
    elif input[i] == "pop":
        result.append(obj.pop())
    elif input[i] == "empty":
        result.append(obj.empty())

print(result)
